File: app/src/main/java/com/example/pam_app/repository/LanguagesRepositoryImpl.java
package com.example.pam_app.repository;

import android.content.SharedPreferences;

import java.util.Locale;

public class LanguagesRepositoryImpl implements LanguagesRepository {

    public static final String KEY_PREF_LANGUAGE = "pref_language";
    private final SharedPreferences sharedPreferences;
    private SharedPreferences.OnSharedPreferenceChangeListener listener;

    public LanguagesRepositoryImpl(final SharedPreferences sharedPreferences) {
        this.sharedPreferences = sharedPreferences;
    }

    @Override
    public Locale getCurrentLocale() {
        return new Locale(sharedPreferences.getString(KEY_PREF_LANGUAGE, "en"));
    }

    @Override
    public String getKeyPrefLanguage(final String key) {
        return sharedPreferences.getString(key, "");
    }

    @Override
    public void setOnSharedPreferencesListener(final SharedPreferences.OnSharedPreferenceChangeListener listener) {
        this.listener = listener;
        this.sharedPreferences.registerOnSharedPreferenceChangeListener(listener);
    }

    @Override
    public void unregisterOnSharedPreferencesListener() {
        sharedPreferences.unregisterOnSharedPreferenceChangeListener(listener);
    }

    @Override
    public void changeLanguage(final String language) {
        sharedPreferences.edit().putString(KEY_PREF_LANGUAGE, language).apply();
    }
}

File: app/src/main/java/com/example/pam_app/repository/LanguagesRepositoryImpl.kt
package com.example.pam_app.repository

import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import java.util.*

class LanguagesRepositoryImpl(private val sharedPreferences: SharedPreferences) : LanguagesRepository {
    private var listener: OnSharedPreferenceChangeListener? = null
    @Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
    override var currentLocale: Locale = Locale(sharedPreferences.getString(KEY_PREF_LANGUAGE, "en"), sharedPreferences.getString(KEY_PREF_COUNTRY, "US"))

    override fun getKeyPrefLanguage(key: String?): String? {
        return sharedPreferences.getString(key, "")
    }

    override fun setOnSharedPreferencesListener(listener: OnSharedPreferenceChangeListener?) {
        this.listener = listener
        sharedPreferences.registerOnSharedPreferenceChangeListener(listener)
    }

    override fun unregisterOnSharedPreferencesListener() {
        sharedPreferences.unregisterOnSharedPreferenceChangeListener(listener)
    }

    override fun changeLanguage(language: String?) {
        sharedPreferences.edit().putString(KEY_PREF_LANGUAGE, language).apply()
        sharedPreferences.edit().putString(KEY_PREF_COUNTRY, if (language == "es") "ES" else "US").apply()
    }

    companion object {
        const val KEY_PREF_LANGUAGE = "pref_language"
        const val KEY_PREF_COUNTRY = "pref_language"
    }

}